src/nbl/ext/ImGui/ImGui.cpp

```diff
@@ -23,61 +20,199 @@ using namespace nbl::ui;
+					const std::string_view code (reinterpret_cast<const char*>(shader->getContent()->getPointer()), shader->getContent()->getSize());
+
+					if constexpr (stage == IShader::E_SHADER_STAGE::ESS_VERTEX)
+					{
+						const auto VERTEX_COMPILE_OPTIONS = std::to_array<std::string>({NBL_DEFAULT_OPTIONS, "-T", "vs_6_7", "-E", "VSMain", "-O3"});
+						options.dxcOptions = VERTEX_COMPILE_OPTIONS;
+
+						return compiler->compileToSPIRV(code.data(), options); // we good here - no code patching
+					}
+					else if (stage == IShader::E_SHADER_STAGE::ESS_FRAGMENT)
+					{
+						const auto FRAGMENT_COMPILE_OPTIONS = std::to_array<std::string>({NBL_DEFAULT_OPTIONS, "-T", "ps_6_7", "-E", "PSMain", "-O3"});
+						options.dxcOptions = FRAGMENT_COMPILE_OPTIONS;
+
+						std::stringstream stream;
+
+						stream << "// -> this code has been autogenerated with Nabla ImGUI extension\n"
+							<< "#define NBL_TEXTURES_BINDING_IX " << m_creationParams.resources.textures.bindingIx << "\n"
+							<< "#define NBL_SAMPLER_STATES_BINDING_IX " << m_creationParams.resources.samplers.bindingIx << "\n"
+							<< "#define NBL_TEXTURES_SET_IX " << m_creationParams.resources.textures.setIx << "\n"
+							<< "#define NBL_SAMPLER_STATES_SET_IX " << m_creationParams.resources.samplers.setIx << "\n"
+							<< "#define NBL_RESOURCES_COUNT " << m_creationParams.resources.count << "\n"
+							<< "// <-\n\n";
+
+						const auto newCode = stream.str() + std::string(code);
+						return compiler->compileToSPIRV(newCode.c_str(), options); // but here we do patch the code with additional define directives for which values are taken from the creation parameters
+					}
+					else
+					{
+						static_assert(stage != IShader::E_SHADER_STAGE::ESS_UNKNOWN, "Unknown shader stage!");
+						return nullptr;
+					}
+				};
+
+				auto spirv = compileToSPIRV();
+
+				if (!spirv)
+				{
+					m_creationParams.utilities->getLogger()->log("Could not compile \"%s\" shader!", system::ILogger::ELL_ERROR, key.value);
+					return nullptr;
+				}
+
+				auto gpu = m_creationParams.utilities->getLogicalDevice()->createShader(spirv.get());
+
+				if (!gpu)
+					m_creationParams.utilities->getLogger()->log("Could not create GPU shader for \"%s\"!", system::ILogger::ELL_ERROR, key.value);
+
+				return gpu;
 			};
 
-			shaders.vertex = createShader(spirv.vertex, IShader::E_SHADER_STAGE::ESS_VERTEX);
-			shaders.fragment = createShader(spirv.fragment, IShader::E_SHADER_STAGE::ESS_FRAGMENT);
+			system->mount(smart_refctd_ptr(archive), NBL_ARCHIVE_ALIAS.data());
+			shaders.vertex = createShader.template operator() < NBL_CORE_UNIQUE_STRING_LITERAL_TYPE("vertex.hlsl"), IShader::E_SHADER_STAGE::ESS_VERTEX > ();
+			shaders.fragment = createShader.template operator() < NBL_CORE_UNIQUE_STRING_LITERAL_TYPE("fragment.hlsl"), IShader::E_SHADER_STAGE::ESS_FRAGMENT > ();
+			system->unmount(archive.get(), NBL_ARCHIVE_ALIAS.data());
```

Comment:
where is `NBL_ARCHIVE_ALIAS` defined?

P.S. maybe we should have a `static` method in the extension to mount the archive in-case the user wants to make their own custom shaders but with the shared headers from the archive?

Reply:
> P.S. maybe we should have a static method in the extension to mount the archive in-case the user wants to make their own custom shaders but with the shared headers from the archive?

fine, can be done

Reply:
resolved